### PR TITLE
Add a way for tooling to determine if a span corresponds to a TagHelper.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- This enables the editor to ask us if an HTML parent happens to be a TagHelper.

#1778

### CODE REVIEW

File: src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    public static class RazorSyntaxFactsServiceExtensions
+    {
+        public static bool IsTagHelperSpan(this RazorSyntaxFactsService syntaxFactsService, SourceSpan span, RazorSyntaxTree syntaxTree)
```
Comment (ajaybhargavb) [inline src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: Why an extension? Why not make it part of `RazorSyntaxFactsService`?

File: src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            if (syntaxFactsService == null)
+            {
+                throw new System.ArgumentNullException(nameof(syntaxFactsService));
```
Comment (ajaybhargavb) [inline src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: using System;

File: src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            }
+
+            if (syntaxTree == null)
+            {
+                return false;
```
Comment (ajaybhargavb) [inline src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: Why not throw? That is what we do in `GetTagHelperSpans` and `GetClassifiedSpans`

Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: Also null-check `span`.

Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: > Why not throw? That is what we do in GetTagHelperSpans and GetClassifiedSpans

> Why not throw? That is what we do in GetTagHelperSpans and GetClassifiedSpans

It is kind of icky to not throw for a null syntax tree but it's possible that the tooling side of the equation doesn't yet have a SyntaxTree. This is just being extra protective.



> Also null-check span.

It's a struct, can't be `null`.

Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: > This is just being extra protective.

Okay. I suggest leaving a comment then.

Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: > It's a struct, can't be null.

💯 PranavPoints will be awarded if you still manage to null check this.  😆

File: src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs (hunk 1 above)
Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: I thought about putting it on the service directly but didn't for the reasons:
1. This method utilizes the services `GetTagHelperSpans` public implementation. Basically feels weird for a helper method to sit on a public contract.
2. It's totally different than the rest of the `RazorSyntaxFactsService` APIs.

Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: > This method utilizes the services GetTagHelperSpans public implementation. Basically feels weird for a helper method to sit on a public contract.

Correct me if I am wrong, but I don't think we should be concerned about this. They are both just implementation details. If someone re-implements this, they have to re-implement all the methods.

> It's totally different than the rest of the RazorSyntaxFactsService APIs.

It looks pretty similar to me but I don't have a strong opinion on this. If you think this should be an extension just for this reason, I think that's fine.

Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: > Correct me if I am wrong, but I don't think we should be concerned about this. They are both just implementation details. If someone re-implements this, they have to re-implement all the methods.

As an extension method that depends on the `GetTagHelperSpans` API they don't need to/can't re-implement `IsTagHelperSpan` because it acts off of the output of the `GetTagHelperSpans` method. Basically it works as long as `GetTagHelperSpans` works 😉

Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: Alright. That makes sense. 👍

File: test/Microsoft.VisualStudio.Editor.Razor.Test/RazorSyntaxFactsServiceExtensionsTest.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            var service = new DefaultRazorSyntaxFactsService();
+
+            // Act
+            var result = service.IsTagHelperSpan(location, syntaxTree);
+
+            // Assert
+            Assert.True(result);
+        }
+        [Fact]
+        public void IsTagHelperSpan_ReturnsFalse()
```
Comment (ajaybhargavb) [inline test/Microsoft.VisualStudio.Editor.Razor.Test/RazorSyntaxFactsServiceExtensionsTest.cs]: Add a test for null syntax tree case.

File: src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs (hunk 1 above)
Comment (ajaybhargavb) [reply src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxFactsServiceExtensions.cs]: nit: Put `syntaxTree` before `span` in the parameters.
